### PR TITLE
[SPARK-12164] [SQL] Decode the encoded values and then display

### PR DESCRIPTION
**Pull request by gatorsmile**
Based on the suggestions from @marmbrus @cloud-fan in https://github.com/apache/spark/pull/10165 , this PR is to print the decoded values(user objects) in `Dataset.show`

``` scala
    implicit val kryoEncoder = Encoders.kryo[KryoClassData]
    val ds = Seq(KryoClassData("a", 1), KryoClassData("b", 2), KryoClassData("c", 3)).toDS()
    ds.show(20, false);
```

The current output is like 

```
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|value                                                                                                                                                                                 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|[1, 0, 111, 114, 103, 46, 97, 112, 97, 99, 104, 101, 46, 115, 112, 97, 114, 107, 46, 115, 113, 108, 46, 75, 114, 121, 111, 67, 108, 97, 115, 115, 68, 97, 116, -31, 1, 1, -126, 97, 2]|
|[1, 0, 111, 114, 103, 46, 97, 112, 97, 99, 104, 101, 46, 115, 112, 97, 114, 107, 46, 115, 113, 108, 46, 75, 114, 121, 111, 67, 108, 97, 115, 115, 68, 97, 116, -31, 1, 1, -126, 98, 4]|
|[1, 0, 111, 114, 103, 46, 97, 112, 97, 99, 104, 101, 46, 115, 112, 97, 114, 107, 46, 115, 113, 108, 46, 75, 114, 121, 111, 67, 108, 97, 115, 115, 68, 97, 116, -31, 1, 1, -126, 99, 6]|
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

After the fix, it will be like the below if and only if the users override the `toString` function in the class `KryoClassData`  

``` scala
override def toString: String = s"KryoClassData($a, $b)"
```

```
+-------------------+
|value              |
+-------------------+
|KryoClassData(a, 1)|
|KryoClassData(b, 2)|
|KryoClassData(c, 3)|
+-------------------+
```

If users do not override the `toString` function, the results will be like 

```
+---------------------------------------+
|value                                  |
+---------------------------------------+
|org.apache.spark.sql.KryoClassData@68ef|
|org.apache.spark.sql.KryoClassData@6915|
|org.apache.spark.sql.KryoClassData@693b|
+---------------------------------------+
```

Question: Should we add another optional parameter in the function `show`? It will decide if the function `show` will display the hex values or the object values?

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -161,17 +161,15 @@ class DataFrame private[sql](
...
   /**
-   * Compose the string representing rows for output
-   * @param _numRows Number of rows to show
-   * @param truncate Whether truncate long strings and align cells right
-   */
-  private[sql] def showString(_numRows: Int, truncate: Boolean = true): String = {
+    * Compose the string representing rows for output
+    * @param _numRows Number of rows to show
+    * @param truncate Whether truncate long strings and align cells right
+    */
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Please revert these white space changes (I hate IDEA 15 too...)

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Yeah. Just upgraded to 15.0.2 The problem still exists...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala [hunk 2]
```diff
@@ -42,4 +43,67 @@ private[sql] trait Queryable {
...
+
+  /**
+    * Format the string representing rows for output
+    * @param rows The rows to show
+    * @param numRows Number of rows to show
+    * @param hasMoreData Whether some rows are not shown due to the limit
+    * @param truncate Whether truncate long strings and align cells right
+    *
+    */
+  private[sql] def formatString (rows: Seq[Seq[String]],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala]: This indenting is wrong. https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Indentation

4 space for function args that wrap.

**Comment by marmbrus** [comment]
Can you add tests too?  Probably in `DataFrameSuite`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala]: Sorry, this is caused by IntelliJ IDEA 15. It automatically changes indenting if I copied and pasted. :( Will correct it.

**Comment by gatorsmile** [comment]
Sure, will do. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -225,7 +225,41 @@ class Dataset[T] private[sql](
...
-  def show(numRows: Int, truncate: Boolean): Unit = toDF().show(numRows, truncate)
+  // scalastyle:off println
+  def show(numRows: Int, truncate: Boolean): Unit = println(showString(numRows, truncate))
+  // scalastyle:on println
+
+  /**
+    * Compose the string representing rows for output
+    * @param _numRows Number of rows to show
+    * @param truncate Whether truncate long strings and align cells right
+    */
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: we should use javadoc style instead of scaladoc.  i.e.
```
/**
 *
 *
 */
```

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Yeah. Will correct all these. Thanks!

**Comment by gatorsmile** [comment]
To simplify the possible code merge, I combined the code changes in another related PR: https://github.com/apache/spark/pull/10165. Thank you!

**Comment by 3ourroom** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-12164] [SQL] Decode the encoded values and then display (#10215)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -698,47 +733,47 @@ class Dataset[T] private[sql](
...
+   */
   def unpersist(blocking: Boolean): this.type = {
     sqlContext.cacheManager.tryUncacheQuery(this, blocking)
     this
   }
 
   /**
-    * Mark the [[Dataset]] as non-persistent, and remove all blocks for it from memory and disk.
-    * @since 1.6.0
-    */
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Please undo these changes.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: So sorry for introducing these inconsistency. Undo the changes.

**Comment by marmbrus** [comment]
Thanks, merging to master.
